catalogd/internal/storage/localdir.go

```diff
@@ -20,95 +26,266 @@ import (
 	if err != nil {
-		_ = file.Close()
-		return nil, err
+		return err
 	}
-	if !stat.Mode().IsRegular() {
-		_ = file.Close()
-		return nil, os.ErrNotExist
+
+	f, err := os.Create(catalogIndexFilePath(catalogDir))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	enc := json.NewEncoder(f)
+	enc.SetEscapeHTML(false)
+	return enc.Encode(idx)
+}
+
+func (s *LocalDirV1) BaseURL(catalog string) string {
+	return s.RootURL.JoinPath(catalog).String()
+}
+
+func (s *LocalDirV1) StorageServerHandler() http.Handler {
+	mux := http.NewServeMux()
+
+	mux.HandleFunc(s.RootURL.JoinPath("{catalog}", "api", "v1", "all").Path, s.handleV1All)
+	if s.EnableQueryHandler {
+		mux.HandleFunc(s.RootURL.JoinPath("{catalog}", "api", "v1", "query").Path, s.handleV1Query)
+	}
+	return mux
+}
+
+func (s *LocalDirV1) handleV1All(w http.ResponseWriter, r *http.Request) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalog := r.PathValue("catalog")
+	catalogFile, catalogStat, err := s.catalogData(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	serveJsonLines(w, r, catalogStat.ModTime(), catalogFile)
+}
+
+func (s *LocalDirV1) handleV1Query(w http.ResponseWriter, r *http.Request) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalog := r.PathValue("catalog")
+	catalogFile, catalogStat, err := s.catalogData(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	defer catalogFile.Close()
+
+	schema := r.URL.Query().Get("schema")
+	pkg := r.URL.Query().Get("package")
+	name := r.URL.Query().Get("name")
+
+	if schema == "" && pkg == "" && name == "" {
+		// If no parameters are provided, return the entire catalog (this is the same as /api/v1/all)
+		serveJsonLines(w, r, catalogStat.ModTime(), catalogFile)
+		return
+	}
+	idx, err := s.getIndex(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	indexReader, ok := idx.Get(catalogFile, schema, pkg, name)
+	if !ok {
+		httpError(w, fs.ErrNotExist)
+		return
+	}
+	serveJsonLines(w, r, catalogStat.ModTime(), indexReader)
+}
+
+func (s *LocalDirV1) catalogData(catalog string) (*os.File, os.FileInfo, error) {
+	catalogFile, err := os.Open(catalogFilePath(s.catalogDir(catalog)))
+	if err != nil {
+		return nil, nil, err
+	}
+	catalogFileStat, err := catalogFile.Stat()
+	if err != nil {
+		return nil, nil, err
+	}
+	return catalogFile, catalogFileStat, nil
+}
+
+func httpError(w http.ResponseWriter, err error) {
+	var code int
+	switch {
+	case errors.Is(err, fs.ErrNotExist):
+		code = http.StatusNotFound
+	case errors.Is(err, fs.ErrPermission):
+		code = http.StatusForbidden
+	case errors.Is(err, ErrInvalidParams):
+		code = http.StatusBadRequest
+	default:
+		code = http.StatusInternalServerError
+	}
+	http.Error(w, fmt.Sprintf("%d %s", code, http.StatusText(code)), code)
+}
+
+func serveJsonLines(w http.ResponseWriter, r *http.Request, modTime time.Time, rs io.ReadSeeker) {
+	w.Header().Add("Content-Type", "application/jsonl")
+	http.ServeContent(w, r, "", modTime, rs)
+}
```

Comment:
With my suggestion of backing away from the multi-read-seeker code, we'll have to have a different version of this function that the query handler can use (the reader returned by the index will no longer be a io.ReadSeeker, it'll just be a reader)

Unfortunately, if we don't use `http.ServeContent` for the query handler, we also do not get the `Last-Modified`/`If-Modified-Since` header handling. Does the standard library have a way for us to implement this caching logic so that we avoid implementing bugs in our own implementation? If we copy code from the Go stdlib, let's just make sure it's license allows that and we do whatever attribution is necessary.

Reply:
Was able to find a way to `http.ServeContent` with `io.Reader` https://github.com/operator-framework/operator-controller/pull/1643/commits/79da4de47b2c40799b9fcea7dec481faf62b04b0

Reply:
Copied the `checkPreconditions` function and all the code associated with it from the net/http package in the `http_preconditions_check.go` file. I did have to make modifications to the copied code too. Imho it's smells funky to me, it's a lot of copied code that we'll have to hope won't carry any bugs, especially with the modifications I did. But, for now it's in there.